tests/responses/test_fileresponse.py

```diff
@@ -9,31 +10,166 @@
+            def name(self):
+                return self._file.name
+
+            def close(self):
+                if self._file:
+                    self._file.close()
+                    self._file = None
+
+            def __enter__(self):
+                return self
+
+            def __exit__(self, e_type, e_val, e_tb):
+                self.close()
+
+        file = TestFile(__file__, 'rb')
+        file.seek(10)
+        response = FileResponse(file)
+        self.assertEqual(response.headers['Content-Length'], str(os.path.getsize(__file__) - 10))
+        response.close()
+
+    def test_content_type_file(self):
+        response = FileResponse(open(__file__, 'rb'))
+        self.assertIn(response.headers['Content-Type'], ['text/x-python', 'text/plain'])
+        response.close()
+
+    def test_content_type_buffer(self):
+        response = FileResponse(io.BytesIO(b'binary content'))
         self.assertEqual(response.headers['Content-Type'], 'application/octet-stream')
+
+    def test_content_type_buffer_explicit(self):
+        response = FileResponse(io.BytesIO(b'binary content'), content_type='video/webm')
+        self.assertEqual(response.headers['Content-Type'], 'video/webm')
+
+    def test_content_type_buffer_explicit_default(self):
+        response = FileResponse(io.BytesIO(b'binary content'), content_type='text/html')
+        self.assertEqual(response.headers['Content-Type'], 'text/html')
+
+    def test_content_type_buffer_named(self):
+        test_tuples = (
+            (__file__, ['text/x-python', 'text/plain']),
+            (__file__ + 'nosuchfile', ['application/octet-stream']),
+            ('test_fileresponse.py', ['text/x-python', 'text/plain']),
+            ('test_fileresponse.pynosuchfile', ['application/octet-stream']),
+        )
+        for filename, content_types in test_tuples:
+            with self.subTest(filename=filename):
+                buffer = io.BytesIO(b'binary content')
+                buffer.name = filename
+                response = FileResponse(buffer)
+                self.assertIn(response.headers['Content-Type'], content_types)
+
+    def test_content_disposition_file(self):
+        filenames = (
+            ('', 'test_fileresponse.py'),
+            ('custom_name.py', 'custom_name.py'),
+        )
+        dispositions = (
+            (False, 'inline'),
+            (True, 'attachment'),
+        )
+        for (filename, header_filename), (as_attachment, header_disposition)\
+                in itertools.product(filenames, dispositions):
+            with self.subTest(filename=filename, disposition=header_disposition):
+                response = FileResponse(open(__file__, 'rb'), filename=filename, as_attachment=as_attachment)
+                self.assertEqual(
+                    response.headers['Content-Disposition'],
+                    '%s; filename="%s"' % (header_disposition, header_filename),
+                )
+                response.close()
```

Comment:
doesn't it apply to all file-opening tests, and not just the couple marked?
also, while it would help with the assertion fails, it still doesn't help in the case of an exception while creating the FileResponse object.
so basically we'd have to surround every test with something like this:
```
file = open(...)
try:
    resp = FileResponse(file, ...)
    resp.close()
    # assertions etc. here
except Exception:
    raise
else:
    file.close()
```
is that the way to go?

Reply:
I guess first thing would be to verify that that we can just do:
```
file = open('whatever')
resp = FileResponse(file)
resp.close()
# assert on the resp object that's already closed
```
Is that so? It appeared at a quick glance to me, but you've spent vastly more time thinking about this!

Reply:
If more of the tests use temp files instead of the actual test module file itself then I think we can avoid this super-duper safety to wrap the actual file creation.

We're hoping the assertions fail someday and catch a bug (or, barring that, we're hoping Carlton downloads your patch and verifies your tests in commit 2 fail without the fixes in commit 3, and I don't want to bust Carlton's setup!)

Reply:
> I guess first thing would be to verify that that we can just do:
> 
> ```
> file = open('whatever')
> resp = FileResponse(file)
> resp.close()
> # assert on the resp object that's already closed
> ```
> 
> Is that so? It appeared at a quick glance to me, but you've spent vastly more time thinking about this!
[i mean, it does happen in a part that i haven't touched, so...](https://github.com/django/django/blob/981acbaf38bfd1964ab0d72d4997e57bd7d037f5/tests/responses/test_fileresponse.py#L185)
`resp.close()` doesn't "close" the object (what would it even mean?), it just runs the `.close()` methods of opened files.

Reply:
As I added below, I'll avoid suggesting things that would increase scope -- but for the very quick win to have response.close() called before the assertions.